README.md
# yumserver Cookbook
[![Build Status](https://travis-ci.org/kemra102/yumserver-cookbook.svg?branch=master)](https://travis-ci.org/kemra102/yumserver-cookbook)

#### Table of Contents

1. [Overview](#overview)
2. [Requirements](#requirements)
3. [Attributes](#attributes)
4. [Usage](#usage)
    * [yumserver_mirror](#yumserver_mirror)
    * [yumserver_rsync_mirror](#yumserver_rsync_mirror)
5. [Contributing](#contributing)
6. [License & Authors](#license-and-authors)

## Overview

This cookbook allows you to mirror remote Yum repos and serve them up via NGINX.

## Requirements

Requires Chef 12.5 or later as this cookbook makes use of [Custom Resources](https://www.chef.io/blog/2015/10/08/chef-client-12-5-released/).

## Attributes

### yumserver::default
| Key                       | Type    | Description                                | Default          |
|:-------------------------:|:-------:|:------------------------------------------:|:----------------:|
| ['yumserver']['basepath'] | String  | Where to store mirroed yum repos.          | /var/lib/yum-rep |

## Usage

You always need to include the main recipe:

```ruby
include_recipe 'yumserver::default'
```

This creates the `basepath` and installs the following packages:

* yum-utils
* createrepo
* rsync

Additionally the recipe calls `yumserver::_nginx` which;

* Sets-up the upstream NGINX repo.
* Installs `nginx`.
* Configures `nginx` to serve up the `basepath`.
* Manages the `nginx` service.

### yumserver_mirror

Each Yum repo you wish to mirror can be defined using the `yumserver_mirror` custom resource.

Each `yumserver_mirror` has the following attributes:

| Attribute        | Type             | Description                                              | Default           |
|:----------------:|:----------------:|:--------------------------------------------------------:|:-----------------:|
| name             | String or Symbol | Resource name.                                           | N/A               |
| local_path       | String           | The basepath where the repo should be stored.            | /var/lib/yum-repo |
| repo_name        | String           | Name of the Yum repo.                                    | N/A               |
| repo_description | String           | Description of the Yum repo.                             | N/A               |
| repo_baseurl     | String           | Base URL of the Yum repo.                                | N/A               |
| options          | String           | Additional options to pass to `reposync`.                | N/A               |
| timeout          | Integer          | Timeout for the `reposync` command.                      | 600               |
| use_repo         | Boolean          | If the repo should be availble for the yumserver to use. | true              |
| repo_workers     | Integer          | Number of createrepo workers to start up.                | server cpu count  |

To Mirror EPEL for EL7 for example:

```ruby
yumserver_mirror 'epel7' do
  repo_name 'epel'
  repo_description 'Extra Packages for Enterprise Linux 7'
  repo_baseurl 'https://dl.fedoraproject.org/pub/epel/7/x86_64/'
  action :create
end
```

### yumserver_rsync_mirror

Each Yum repo you wish to mirror that provides an rsync endpoint can be defined using the `yumserver_rsync_mirror` custom resource.

Each `yumserver_rsync_mirror` has the following attributes:

| Attribute        | Type             | Description                                              | Default                                                    |
|:----------------:|:----------------:|:--------------------------------------------------------:|:----------------------------------------------------------:|
| name             | String or Symbol | Resource name.                                           | N/A                                                        |
| local_path       | String           | The basepath where the repo should be stored.            | /var/lib/yum-repo                                          |
| repo_name        | String           | Name of the Yum repo.                                    | N/A                                                        |
| repo_description | String           | Description of the Yum repo.                             | N/A                                                        |
| repo_url         | String           | URL of the Yum repo.                                     | N/A                                                        |
| timeout          | Integer          | Timeout for the `rsync` command.                         | 600                                                        |
| rsync_options    | String           | Options to pass to rsync.                                | -aHS --numeric-ids --delete --delete-delay --delay-updates |
| use_repo         | Boolean          | If the repo should be availble for the yumserver to use. | true                                                       |

To Mirror EPEL for EL7 for example:

```ruby
yumserver_rsync_mirror 'epel7' do
  repo_name 'epel'
  repo_description 'Extra Packages for Enterprise Linux 7'
  repo_url 'rsync://dl.fedoraproject.org/pub/epel/7/x86_64/'
  action :create
end
```

## Contributing

If you would like to contribute to this cookbook please follow these steps;

1. Fork the repository on Github.
2. Create a named feature branch (like `add_component_x`).
3. Write your change.
4. Write tests for your change (if applicable).
5. Run the tests, ensuring they all pass.
6. Submit a Pull Request using Github.

## License and Authors

License: [BSD 2-clause](https://tldrlegal.com/license/bsd-2-clause-license-\(freebsd\))

Authors:

  * [Danny Roberts](https://github.com/kemra102)
  * [All Contributors](https://github.com/kemra102/yumserver-cookbook/graphs/contributors)


CHANGELOG.md
## 2017-01-05 - Release 2.1.0
### Summary
Additional features for custom resources.

#### Features
- Added a `timeout` option for both `yumserver_mirror` & `yumserver_rsync_mirror`.
- Added an `options` option for `yumserver_mirror` allowing you to add additional options to the reposync command.

## 2016-02-14
### Summary
Minor bug fix.

#### Bugfixes
- Fixed a bug that meant the `node['yumserver']['basepath']` attribute wasn't honoured by the `yumserver_mirror` & `yumserver_rsync_mirror` resources.

## 2016-02-11
### Summary
Minor bug fix after a fault was found in production with this cookbook.

#### Bugfixes
- The path defined by the `yumserver_mirror` resource was incorrect. This was causing the `zap` feature to fail on runs after the first. Fixing the defined path in this resource fixed the error with `zap`.

## 2016-02-09
### Summary
Major release with additional features/bug fixes that may break backwards compatibility.

### Features
- Now using `mixlib-shellout` instead of exec resources for `yumserver_mirrror` resources.
- The `createrepo` command ran as part of `yumserver_mirror` resources are now multi-threaded.
- Add `zap` support for `/etc/resposync.repos.d`.
- Now uses the `nginx_server` cookbook to provide the web server.

### Bugfixes
- Fixed a typo in the `yumserver_mirror` resource that prevented the reposync config from beign remove on a `:delete` action.

## 2015-12-02
### Summary
Small bugfix update to make the delete action work properly across custom resources.

#### Bugfixes
- In the mirror resource make sure the reposync file gets deleted by switching the resource type to file instead of template.
- In both resources make sure the base path is deleted recursively (otherwise it will fail as there are packages in them).

## 2015-11-25 1.3.0
### Summary
Large bugix release to ensure mirrored repos are properly updated after the 1st sync.

#### Bugfixes
- Made `reposync` in the `mirror` resource use alternate yum.conf & repo directory to allow them all to be enabled (enabling updates) whithout interfering with the system's yum.
- Made the metadata on `reposync` in the `mirror` resource expire immedietly so package changes are synced more often than every 6 hours.

## 2015-11-22 - Release 1.2.0
### Summary
General improvements to mirror resources.

#### Features
- Optionally (`true` by default) create a `yum_repository` to allow the yumserver to use it's mirrored repos.
- The `yumserver_mirror` resource now correctly removes packages no longer in the upstream repo.
- The `yumserver_mirror` should complete quicker when no packages are updated.

#### Bugfixes
- Disable mirrored repos to avoid issues such as mirroring the same repo for two versions of the same distro and causing update issues on the yumserver.

## 2015-11-12 - Release 1.1.1
### Summary
Let Yum repo servers use their own rsync mirrors as repos.

#### Features
- For each `yumserver_rsync_mirror` create a `yum_repository` so the Yum server can use that repo if it needs it.

## 2015-11-10 - Release 1.1.0
### Summary
Added the ability to mirror a repo that is available via rsync.

#### Features
- Mirror yum repos that provide rsync as an option.

#### Bugfixes
- Add regression test.

## 2015-11-10 - Release 1.0.2
### Summary
Fix NGINX config.

#### Bugfixes
- [NGINX recipe does not configure basepath](https://github.com/kemra102/yumserver-cookbook/issues/1)

## 2015-10-11 - Release 1.0.1
### Summary
Minor doc updates.

## 2015-10-11 - Release 1.0.0
### Summary
Initial release. Includes basic functionality.

#### Features
- Mirrors Yum repos.
- Serves Yum repos via NGINX.


test/fixtures/cookbooks/yumserver-test/metadata.rb
name 'yumserver-test'
version '1.0.0'

depends 'yumserver'


test/integration/default/serverspec/yumserver_spec.rb
require 'serverspec'

set :backend, :exec

%w(yum-utils createrepo rsync nginx).each do |pkg|
  describe package(pkg) do
    it { should be_installed }
  end
end

describe file('/etc/reposync.conf') do
  it { should be_file }
  it { should be_mode 644 }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
end

%w(/etc/reposync.repos.d /yum /yum/nginx /yum/nginx/repodata).each do |dir|
  describe file(dir) do
    it { should be_directory }
    it { should be_mode 755 }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
  end
end

describe file('/etc/reposync.repos.d/nginx.repo') do
  it { should be_file }
  it { should be_mode 644 }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
end

%w(/yum/centos-xen /yum/centos-xen/repodata).each do |dir|
  describe file(dir) do
    it { should be_directory }
  end
end

describe file('/etc/nginx/conf.d/yumserver.conf') do
  it { should be_file }
  it { should be_mode 644 }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should contain 'root /yum;' }
end

describe service('nginx') do
  it { should be_enabled }
  it { should be_running }
end

describe port(80) do
  it { should be_listening }
end


test/fixtures/cookbooks/yumserver-test/recipes/default.rb
include_recipe 'yumserver'

major_ver = node['platform_version']

yumserver_mirror 'nginx' do
  repo_name 'nginx'
  repo_description 'NGINX web server'
  repo_baseurl "http://nginx.org/packages/centos/#{major_ver[/./,0]}/x86_64/"
  options '-g'
  timeout 800
  action :create
end

yumserver_rsync_mirror 'centos-xen' do
  repo_name 'centos-xen'
  repo_description 'CentOS Virtualization Packages'
  repo_url "rsync://mirror.bytemark.co.uk/centos/#{major_ver[/./,0]}/virt/x86_64/xen/"
  action :create
end


resources/mirror.rb
property :name, [String, Symbol], required: true, name_property: true
property :local_path, String, required: false
property :repo_name, String, required: true
property :repo_description, String, required: true
property :repo_baseurl, String, required: true
property :options, String, required: false, default: ''
property :timeout, Integer, required: false, default: 600
property :use_repo, [TrueClass, FalseClass], required: true, default: true
property :repo_workers, Integer, required: false, default: node['cpu']['total']

def real_local_path
  if local_path == NilClass
    "#{local_path}/#{name}/"
  else
    "#{node['yumserver']['basepath']}/#{name}/"
  end
end

# Needs to be def'd as path for zap to work.
def path
  "/etc/reposync.repos.d/#{repo_name}.repo"
end

action :create do
  template path do
    cookbook 'yumserver'
    source 'repo.erb'
    owner 'root'
    group 'root'
    mode '0644'
    variables(
      repo_name: repo_name,
      repo_description: repo_description,
      repo_baseurl: repo_baseurl
    )
    action :create
  end
  directory real_local_path do
    owner 'root'
    group 'root'
    mode '0755'
    action :create
  end
  ruby_block 'reposync' do
    block do
      YumServer::Helper.reposync(repo_name, real_local_path, options,
                                 new_resource.timeout)
    end
    action :run
    only_if { ::File.exist?(path) }
  end
  ruby_block 'createrepo' do
    block do
      YumServer::Helper.createrepo(real_local_path, repo_workers)
    end
    action :run
  end
  if use_repo
    yum_repository repo_name do
      description repo_description
      baseurl "file://#{real_local_path}"
      gpgcheck false
      action :create
    end
  end
end

action :delete do
  file path do
    action :delete
  end
  directory real_local_path do
    recursive true
    action :delete
  end
  yum_repository repo_name do
    action :delete
  end
end


resources/rsync_mirror.rb
property :name, [String, Symbol], required: true, name_property: true
property :local_path, String, required: false
property :repo_name, String, required: true
property :repo_description, String, required: true
property :repo_url, String, required: true
property :timeout, Integer, required: false, default: 600
property :rsync_options, String, required: false,
                                 default: '-aHS --numeric-ids --delete \
                                 --delete-delay --delay-updates'
property :use_repo, [TrueClass, FalseClass], required: true, default: true

def real_local_path
  if local_path == NilClass
    "#{local_path}/#{name}/"
  else
    "#{node['yumserver']['basepath']}/#{name}/"
  end
end

def path
  real_local_path
end

action :create do
  directory real_local_path do
    owner 'root'
    group 'root'
    mode '0755'
    action :create
  end
  ruby_block 'rsync' do
    block do
      YumServer::Helper.rsync(rsync_options, repo_url, real_local_path,
                              new_resource.timeout)
    end
  end
  if use_repo
    yum_repository repo_name do
      description repo_description
      baseurl "file://#{real_local_path}"
      gpgcheck false
      action :create
    end
  end
end

action :delete do
  directory real_local_path do
    recursive true
    action :delete
  end
  yum_repository repo_name do
    action :delete
  end
end
